### ISSUE TITLE
Potential segmentation fault in main.cpp in some compiling environments

### ISSUE THREAD

**Issue by WilsonWangTHU** [alecjacobson/computer-graphics-raster-images#15]
Hi guys,

I noticed that for some machines (my personal pc and some U of T clusters),
the original project (no modifications made, fresh cloned) could be compiled, but throws segmentation fault if run.

I am using ```Ubuntu 18.04.1 LTS```, ```cmake 3.10.2```, ```g++ (Ubuntu 7.3.0-16ubuntu3) 7.3.0```, with ```libstdc++.so.6```

```zsh
Program received signal SIGSEGV, Segmentation fault.
__strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:62
62      ../sysdeps/x86_64/multiarch/strlen-avx2.S: No such file or directory.
(gdb) up
#1  0x00007ffff7b7282b in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) up
#2  0x000055555556f0f6 in main (argc=1, argv=0x7fffffffdcf8) at /home/wtw/csc2504/csc418-A1/main.cpp:97
97          read_rgba_from_png(input_filenames[f],next_rgba,next_width,next_height);
```

I suspect it has something to do with c++'s implicit constructor, as the function
 ```inline bool read_rgba_from_png``` asks for
```c++
const std::string & filename
```
while is provided with 
``` c++
input_filenames = const_cast<char **>(default_input_file_names);
...
input_filenames[f]
``` 

After fixing line 18-29 from main.cpp into 
```c++
  const std::string input_filenames[] = {
      "../shared/data/dog.png",
      "../shared/data/glasses.png",
      "../shared/data/laser-beams.png",
      "../shared/data/sparkles.png"
  };
  int num_inputs = 4;

```
it could be compiled and run correctly.

I am not familiar with C++, and I might have made some silly mistakes here.
It will be great if someone could help. And btw, is there a way we could test if our project could be compiled correctly when tested?

Thx in advance ;P

**Comment by risingdhxs** [comment]
I just wanna say that I ran into a similar issue: read_rgba_from_png reads the dog picture fine for the first time, but the second time (for the over part) it got a seg fault:
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)

By inserting couts in read_rgba.h file I see that when it ran the second time it actually went into the if (rgba_raw==NULL) branch.

Wilson's fix didn't work for me though, since CLion (I'm running in Ubuntu 18.04 btw) gave me this:
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid

Please instruct as to what to do.

Thanks.

Gary

**Comment by WilsonWangTHU** [comment]
@risingdhxs Hi Gary,

Since both of us are using Ubuntu 18.04, whose default version of g++ is 4.7.3.*,
It might be worth a trial to switch back to g++ 4.4.* or maybe 4.5.* ;P

**Comment by alecjacobson** [comment]
I can't reproduce this with `c++ (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0`. Is this still an issue?

**Comment by bod11** [comment]
By downgrading to g++ 5 I was able to get rid of the seg fault. So it seems the compiler version is the issue. the gcc compiler seems to be fine, specifically the g++ compiler needed to be downgraded to 5. 

sudo apt install g++-5
then if you want you can set it as an alternative and then swap to it.
sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-5 10
sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-5 20

and to swap:
sudo update-alternatives --config g++

you just input the number that corresponds to the version you want to swap to


 I just removed the other g++ compiler since this portable hard drive is just used for school anyway.

**Comment by WilsonWangTHU** [comment]
@alecjacobson it seems that for people start with a ubuntu 18.04+, it might be a good idea to switch back to the older g++, according to @bod11

**Comment by PerpetualSaltMachine** [comment]
where you able to fix it following the steps I used @disconcision ? I am bod11, this is just an alternate git account if I don't show up as bod11 XD.

**Comment by disconcision** [comment]
> where you able to fix it following the steps I used @disconcision ?

Yes I was, thanks, I should have specified. I just wanted to confirm that it indeed seems to be an issue with the newer gcc version on ubuntu 18.

**Comment by alecjacobson** [comment]
Fixed in 86eab0c7078dcaa967f85baaa784a9e1e67b055c.
